### ISSUE TITLE
Suggestion regarding database structure

### ISSUE THREAD

**Issue by shavitush** [sbpp/sourcebans-pp#367]
## Expected Behavior
Fast, near-instant page loads.

## Current Behavior
Some pages load slowly. Example: bans list (`index.php?p=banlist`).

### **Solution 1:**
---
1. Add the following indexes: (as they're used in `WHERE` and `ON` clauses from within the ban list page)

```sql
ALTER TABLE `sb_bans`
	DROP INDEX `sid`,
	ADD INDEX `sid` (`sid`, `aid`, `type`, `ip`, `created`);

ALTER TABLE `sb_servers`
	ADD INDEX `modid` (`modid`);
```

2. Change the default tables' engine to InnoDB, as right now it's MyISAM. AFAIK there's no downside to using InnoDB in MySQL versions 5.6 or above (or MariaDB equivalent).

Applying both the indexes and the InnoDB change resulted in the page loading in 0.3 seconds on my server rather than 6 seconds. 

### **Solution 2:**
---
Optimize the database querying from within the slow pages (such as the bans list).

The main problematic query in the page I mentioned (bans list) is the following: https://github.com/sbpp/sourcebans-pp/blob/3e76d97b8647c520da0c3f2d93305da7154e0865/web/pages/page.banlist.php#L269-L283

(a similar version of it is repeated 3 times in the page and executed in different scenarios)

**Comment by rumblefrog** [comment]
Neat, there are a lot of problems with legacy Sourceban that we discover as we build on top of it; it's only a matter of time before a complete rewrite.

**Comment by shavitush** [comment]
True. I just posted an example with the bans page. As far as I'm concerned, (even by going through the plugins) there are many cases of `WHERE` and `ON` clauses without the column itself being indexed, I'm sure that SourceBans++ can receive a huge performance boost if this is done correctly!

Obviously, rewriting queries to use less I/O; especially less `JOIN`s and subqueries will be great too. Refactoring places where multiple queries are executed at once would be awesome too.

**Comment by stickz** [comment]
### We need to rewrite some tables. 
- Delete _bans, _demos and _comms tables.
- Rewrite with three new tables as posted bellow.
- There will be 5 joins required with NO subqueries.
- Purge scripts will be significantly lower overhead.

### 1. User table with uid as primary surrogate key.
- Community id needs to be added directly to the user.
- History count needs to be incremented for the user.
- Demo count needs to be incremented for the user.
- Gag, Mute, Silence needs to be incremented for the user.

### 2. Ban table with bid as primary surrogate key.
- BanType char B, G, M, S letters.
- Join the User table with uid key.
- Join the Demo table with did key.
- Join the _servers, _mods, _admins tables as before.

### 3. Demo table with did as primary surrogate key.
- straight forward. filename and origname.

**Comment by rumblefrog** [comment]
Marked `help wanted` + `p: medium` as part of the pre-2.0 backlog grooming pass (#1116). Concrete, well-scoped perf win (the SQL is in the body); no current owner — PRs welcome.
